### PR TITLE
 feat: optimize stacked MoE weight loading — per-expert split before broadcast && support fp32 ssm state && fix load quant

### PR DESCRIPTION
**Pull request by alibaba-miji**
## Summary

优化 stacked MoE checkpoint（如 Qwen3.5-35B-A3B）的权重加载流程，将 NCCL 广播粒度从完整 stacked tensor `[num_experts, H, K]` 降低到单个 expert slice `[H, K]`，显著降低加载期间的 GPU 峰值显存占用。

## Problem

Qwen3.5-35B-A3B 等模型的 MoE 权重以 stacked 格式存储在 checkpoint 中——所有 expert 合并为单个 tensor（例如 `experts.down_proj` shape 为 `[256, H, K]`）。

原有加载流程中，fastsafetensors 读取到完整的 stacked tensor 后，直接通过 NCCL 广播给所有 rank。这意味着在广播期间，**每个 GPU 都需要持有完整的 `[256, H, K]` tensor**，对于 256 个 expert 的模型，这会造成巨大的瞬时显存峰值。

### 修改前流程

image

1. fastsafetensors 加载完整 stacked tensor `[256, H, K]` 到 source rank GPU
2. NCCL 广播完整 `[256, H, K]` tensor 到所有 rank
3. 每个 rank 在 GPU 上持有完整 stacked tensor，然后再做 process/split
4. **峰值显存 = 完整 stacked tensor 大小**

### 修改后流程

image

1. `PerExpertParallelLoader` 在 source rank 上将 stacked tensor 按 expert 维度切片
2. NCCL 逐个广播单个 expert slice `[H, K]`
3. 每个 rank 同一时刻只持有一个 expert 的数据
4. **峰值显存 = 单个 expert slice = 原来的 1/E**

### 为什么能节省显存

关键在于**广播粒度的变化**：

| | 广播单元 | 峰值显存占用 |
|---|---|---|
| **Before** | 完整 `[E, H, K]` tensor | `E × H × K × dtype_size` |
| **After** | 单个 `[H, K]` expert slice | `H × K × dtype_size` |

以 Qwen3.5-35B-A3B 为例（256 experts），广播期间的 GPU 峰值显存降低约 **256 倍**。

原有方案中，即使最终每个 rank 只需要部分 expert（EP 场景），加载期间仍然需要在所有 rank 上分配完整 stacked tensor 的显存来接收 NCCL 广播。新方案将切分前移到广播之前，从根本上消除了这一瓶颈。

### New: `PerExpertParallelLoader`
- 继承 `fastsafetensors.ParallelLoader`，重写 `_consume_single_batch`
- 对 stacked MoE tensor，在 source rank 上按 expert 维度切片后逐个广播
- 非 stacked tensor 走原有 broadcast 路径，无影响

### New: `StackSplitTensorSource`
- 包装 `TensorSource`，透明地将 stacked tensor 按 expert 拆分
- 用于非 fastsafetensors 加载路径（fallback），确保行为一致
- 内置 cache：同一个 stacked tensor 只加载一次

### Modified: `MoeAtomicWeight`
- 新增 `stacked_ckpt_keys` 标志，替代原 `MoeConfig.weight_stack`
- 新增 `_build_split_config`、`_get_expert_weights`、`_expert_key_pattern` 方法
- `_load_raw_tensor` 自动检测 stacked keys 并包装 `StackSplitTensorSource`

### Modified: `ModelLoader`
- 新增 `_build_stacked_key_config`，遍历权重树构建 stacked key 映射
- 传递映射到 `fastsafetensors_weights_iterator`，启用 per-expert 广播

### Cleanup
- 移除 `MoeConfig.weight_stack`、`LoadConfig.use_stack_weight`
- 移除 `sp_moe_neg1`/`sp_moe_w1` 中的 `use_stack_weight` 分支
- 修复 `create_w8a8_fp8_per_block_weight` 中 `LinearAttnAtomicWeight` 的 isinstance 顺序
- 修复 `copy.deepcopy` 无法序列化 C 扩展对象的问题（改为浅拷贝）

### CODE REVIEW

**Comment by alibaba-miji** [comment]
### 1. PerExpertParallelLoader 和 StackSplitTensorSource 缺少单元测试 [P1]

**结论：合理，建议采纳。**

核心新增代码确实无测试覆盖。三个关键组件可独立测试：

- `**StackSplitTensorSource`**：可 mock `TensorSource`，测试 cache 命中/miss、split 正确性、非 stacked key passthrough
- `**_build_stacked_key_config`**：纯函数，构造 `weight_info_list` 即可测试映射正确性
- `**MoeAtomicWeight._build_split_config`**：测试 expert 选择和 key 生成

`PerExpertParallelLoader._broadcast_per_expert` 涉及 NCCL/fastsafetensors 深度耦合，单测成本高，可暂缓（通过集成测试覆盖）。

**改进方案：** 在 `rtp_llm/model_loader/test/` 下新增测试文件 `test_stack_split_tensor_source.py`，包含：

- `StackSplitTensorSource` 的 cache 命中/miss、split 切片正确性、边界 case（`num_experts=1`）
- `_build_stacked_key_config` 的映射构建正确性（含 `merge_fun is not identity` 过滤逻辑）
- `MoeAtomicWeight._build_split_config` 的 per-expert key 格式正确性

---

### 2. load_python_model 默认值从 False 改为 True 是 Breaking Change [P1]

**结论：合理，但本质是"混入无关改动"（与 #7 重叠）。**

忽略这个，这个不重要

---

### 3. MoeAtomicWeight._postprocess 与 MoeWeight._postprocess 的 shuffle 重复执行 [P1]

**结论：合理，确认是真实 bug。**

调用链分析：

```mermaid
flowchart TD
    A["MoeWeight._postprocess"] -->|"shuffle w1/w2/s1/s2"| B["_shuff_moe_weight"]
    A -->|"return super()._postprocess"| C["CompositeWeight._postprocess"]
    C -->|"遍历 sub_weights"| D["MoeAtomicWeight._postprocess"]
    D -->|"再次 shuffle w1/w2/s1/s2"| E["shuffle_moe_weight (第二次!)"]
```



具体路径：

1. `MoeWeight._postprocess`（[ffn_weight.py:476-491](rtp_llm/model_loader/ffn_weight.py)）对 `moe_w1/w2/s1/s2` 执行 `shuffle_moe_weight`
2. 然后 `super()._postprocess` -> `CompositeWeight._postprocess`（[weight_module.py:864-878](rtp_llm/model_loader/weight_module.py)）遍历每个 `sub_weight` 调用 `_postprocess`
3. 新增的 `MoeAtomicWeight._postprocess`（[ffn_weight.py:316-330](rtp_llm/model_loader/ffn_weight.py)）对同样的 key 再次调用 `shuffle_moe_weight`

在 `CudaImpl.shuffle_moe_weight`（[device_impl.py:832+](rtp_llm/device/device_impl.py)）中，shuffle 包含 padding 和 block reordering，**双重执行会损坏权重数据**。在 `GpuImpl` 基类中 shuffle 是恒等函数 (`return x`)，因此在非 CUDA/ROCm 设备上不会表现为 bug，这也解释了为什么可能没在测试中发现。

**改进方案：** 将 shuffle 逻辑统一到 `MoeAtomicWeight._postprocess`（每个原子权重自行负责），删除 `MoeWeight._postprocess` 和 `_shuff_moe_weight` 方法。修改后 `MoeWeight` 直接继承 `CompositeWeight._postprocess`。

```python
# ffn_weight.py - MoeWeight 删除以下方法:
# - _shuff_moe_weight (line 459-474)
# - _postprocess (line 476-491)
# MoeWeight 将直接使用 CompositeWeight._postprocess
```

需注意 `MoeWeight._postprocess` 还处理了 `moe_s1/moe_s2`（scale 权重），这些可能是嵌套在 dict 中的。确认 `MoeAtomicWeight._postprocess` 已覆盖 `W.moe_s1`/`W.moe_s2` 的 shuffle（代码 line 320 确认已覆盖）。

---

### 4. PerExpertParallelLoader 深度耦合 fastsafetensors 内部实现 [P1]

**结论：合理，但短期修复代价高。**

`[_broadcast_per_expert](rtp_llm/model_loader/per_expert_parallel_loader.py)` 直接访问的私有 API：

- `fb._get_rank_lidx(key)` - 获取 tensor 所在 rank 和 loader index
- `fb.rank_loaders[rank][lidx]` - 获取 factory
- `factory.metadata.tensors[key]` - 获取 tensor frame (shape/dtype)
- `factory.tensors[key]` - 获取实际 tensor
- `factory.framework.get_empty_tensor(...)` - 创建空 tensor
- `fb.instantiated[rank][lidx][key]` - 标记已消费
- `factory.free_dev_ptrs()` - 释放设备指针

向上游贡献"per-expert broadcast"需要较大的 API 设计工作。

**改进方案（务实）：**

1. 在文件头部添加注释标注依赖的 fastsafetensors 版本
2. 在 `_broadcast_per_expert` 方法上添加注释列出依赖的内部 API
3. 添加 `__init__` 中的版本检查（`assert fastsafetensors.__version__ == "x.y.z"`）

---

### 5. _consume_single_batch 大量复制父类代码 [P2]

**结论：部分合理，但改造空间有限。**

`_consume_single_batch` 约 60 行中，核心差异只有 `for key in batch.keys` 循环中的 `stacked_key_config` 分支（约 4 行）。但由于 fastsafetensors 是外部库，无法通过 hook/template method 模式减少复制。

**改进方案：** 在方法开头添加注释，标注该方法复制自 `ParallelLoader._consume_single_batch`，并用 `# --- BEGIN CUSTOM LOGIC ---` / `# --- END CUSTOM LOGIC ---` 标记差异区域，便于未来同步上游变更。

---

### 6. StackSplitTensorSource._stacked_cache 无清理机制 [P2]

**结论：低风险，但建议添加注释。**

从代码看，`StackSplitTensorSource` 在 `[MoeAtomicWeight._load_raw_tensor](rtp_llm/model_loader/ffn_weight.py)` 中每次调用新建实例，方法返回后实例即可被 GC。cache 不会跨层/跨 weight 积累。

**改进方案：** 在 `_stacked_cache` 声明处添加注释说明其生命周期设计：

```python
# Cache lives for the duration of a single _load_raw_tensor call.
# Each call creates a new StackSplitTensorSource instance, so the
# cache is automatically garbage-collected after the method returns.
self._stacked_cache: Dict[str, torch.Tensor] = {}
```

---

### 7. PR 混入无关改动 [P2]

**结论：合理。**

忽略

---

### 8. torch.cuda.empty_cache() 在每个 stacked tensor 后调用 [P2]

**结论：合理，且问题比 review 指出的更严重 — 这个调用完全无效，应直接删除而非移动。**

深入分析 `_broadcast_per_expert` 执行完时的 GPU 内存状态：

1. **yielded 出去的 expert 切片张量**（`clone().detach().get_raw()`）— 已通过 `yield` 交给调用者存为模型权重，仍然存活，不可回收。
2. **factory 的底层 GPU buffer (`gbuf`)**  — 仅当该 factory **所有 key** 都被消费完才会触发 `free_dev_ptrs()`。通常一个 safetensors 文件包含大量非 MoE key，处理完一个 stacked key 后几乎不可能触发。**绝大多数情况下仍然存活。**
3. **循环中的临时变量** (`expert_t`) — `get_raw()` 只是返回内部 `real_tensor` 引用（见 `TorchTensor.get_raw` 实现），没有产生额外分配。

因此在 `empty_cache()` 调用点：

- **常见情况**：没有任何 GPU 内存被释放到 caching allocator 的 free list，`empty_cache()` 扫描一遍发现是空的，纯粹浪费时间。
- **极少情况**（恰好触发 `free_dev_ptrs`）：`gbuf` 通过 `caching_allocator_delete` 进入 free list，但 `empty_cache()` 会将其刷回 CUDA driver，**反而有害** — 后续 `clone()` / `get_empty_tensor` 需要分配内存时，留在 free list 可直接复用，刷回 driver 则需走慢速 `cudaMalloc`。

**改进方案：** 直接删除 `_broadcast_per_expert` 末尾的 `torch.cuda.empty_cache()`（[per_expert_parallel_loader.py:121](rtp_llm/model_loader/per_expert_parallel_loader.py)）。不需要移到其他位置。

---

### 9. ssm_state_dtype_str_to_data_type 不支持 fp16 [P3]

**结论：合理。**

`[ssm_state_dtype_str_to_data_type](rtp_llm/config/model_config.py)` 仅支持 `bf16` 和 `fp32`。`typed_storage_view.py` 的 `TORCH_DTYPE_SIZE_BYTES` 已包含 `fp16`，说明底层已有支持。

**改进方案：** 添加 fp16 映射：

```python
def ssm_state_dtype_str_to_data_type(ssm_state_dtype: str) -> DataType:
    ssm_state_dtype = ssm_state_dtype.lower()
    if ssm_state_dtype == "bf16":
        return DataType.TYPE_BF16
    if ssm_state_dtype == "fp16":
        return DataType.TYPE_FP16
    if ssm_state_dtype == "fp32":
        return DataType.TYPE_FP32
    raise ValueError(f"Unsupported ssm_state_dtype: {ssm_state_dtype}")
```

---

### 10. store_ssm_state_to_block_map 的 fp32 硬断言 [P3]

**结论：合理。**

`[store_ssm_state_to_block_map](rtp_llm/models_py/triton_kernels/fla/block.py:192-194)` 硬性要求 `h.dtype == torch.float32 and final_states.dtype == torch.float32`，但没有解释原因。

从 Triton kernel 实现看，`store_ssm_state_to_block_map_kernel` 中使用 `tl.load` 直接读取 `h` 和 `final_states`，kernel 内部的中间计算依赖 fp32 精度。SSM 状态的累积特性使低精度容易产生数值漂移。

**改进方案：** 在 assert 前添加注释说明原因：

```python
# h and final_states require fp32 because the SSM state accumulation
# in the Triton kernel relies on fp32 precision to avoid numerical
# drift across chunks. The kernel loads these tensors directly without
# dtype conversion.
assert (
    h.dtype == torch.float32 and final_states.dtype == torch.float32
), "h and final_states must be float32"
```

---

## 优先级总结


| 编号  | 优先级 | 结论                    | 改动量                       |
| --- | --- | --------------------- | ------------------------- |
| #3  | P1  | **Bug: shuffle 双重执行** | 删除 MoeWeight 的 shuffle 方法 |
| #1  | P1  | 添加单元测试                | 新增测试文件                    |
| #2  | P1  | Breaking change 应拆 PR | 仅 PR 管理                   |
| #4  | P1  | 添加版本检查和注释             | 小改动                       |
| #7  | P2  | 拆分无关改动                | PR 管理                     |
| #8  | P2  | empty_cache 无效，直接删除   | 删除一行代码                    |
| #5  | P2  | 添加差异标注注释              | 注释                        |
| #6  | P2  | 添加 cache 生命周期注释       | 注释                        |
| #9  | P3  | 添加 fp16 支持            | 2 行代码                     |
| #10 | P3  | 添加注释                  | 注释                        |
